jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/CloudWatchLogStreamClient.kt

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.cloudwatch.logs
+
+import com.intellij.openapi.project.Project
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.future.asDeferred
+import software.amazon.awssdk.services.cloudwatchlogs.CloudWatchLogsAsyncClient
+import software.amazon.awssdk.services.cloudwatchlogs.model.GetLogEventsRequest
+import software.amazon.awssdk.services.cloudwatchlogs.model.OutputLogEvent
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import java.util.concurrent.atomic.AtomicBoolean
+
+class CloudWatchLogStreamClient(
+    project: Project,
+    private val logGroup: String,
+    private val logStream: String
+) : CoroutineScope by ApplicationThreadPoolScope("CloudWatchLogsStream") {
+    private val client: CloudWatchLogsAsyncClient = project.awsClient()
+    private var nextBackwardToken: String? = null
+    private var nextForwardToken: String? = null
+    private val loadingForwardEvent: AtomicBoolean = AtomicBoolean(false)
+    private val loadingBackwardEvent: AtomicBoolean = AtomicBoolean(false)
+
+    suspend fun loadInitialAround(startTime: Long, timeScale: Long): List<OutputLogEvent> {
+        try {
+            loadingForwardEvent.set(true)
+            loadingBackwardEvent.set(true)
+            val request = GetLogEventsRequest
+                .builder()
+                .logGroupName(logGroup)
+                .logStreamName(logStream)
+                .startTime(startTime - timeScale)
+                .endTime(startTime + timeScale)
+                .build()
+            return load(request, saveForwardToken = true, saveBackwardToken = true)
+        } finally {
+            loadingBackwardEvent.set(false)
+            loadingForwardEvent.set(false)
+        }
+    }
+
+    suspend fun loadInitial(fromHead: Boolean): List<OutputLogEvent> {
+        try {
+            loadingForwardEvent.set(true)
+            loadingBackwardEvent.set(true)
+            val request = GetLogEventsRequest.builder().logGroupName(logGroup).logStreamName(logStream).startFromHead(fromHead).build()
+            return load(request, saveForwardToken = true, saveBackwardToken = true)
+        } finally {
+            loadingBackwardEvent.set(false)
+            loadingForwardEvent.set(false)
+        }
+    }
+
+    suspend fun loadMoreForward(): List<OutputLogEvent> = loadMore(loadingForwardEvent, nextForwardToken, saveForwardToken = true)
+
+    suspend fun loadMoreBackward(): List<OutputLogEvent> = loadMore(loadingBackwardEvent, nextBackwardToken, saveBackwardToken = true)
+
+    private suspend fun loadMore(
+        loadingGuard: AtomicBoolean,
+        nextToken: String?,
+        saveForwardToken: Boolean = false,
+        saveBackwardToken: Boolean = false
+    ): List<OutputLogEvent> {
+        val alreadyLoading = loadingGuard.getAndSet(true)
```

Comment:
I'm curious if we should bring in a coroutine actor instead of us doing this ourselves.

Actor processes loading more data, we just keep pumping load reqeusts into the channel

Reply:
As long as the actor library can also dedupe, we want to throw out requests while processing otherwise we will request the same bunch of logs twice

Reply:
I updated it using channels